File: spk/cli/_cmd_test.py
# Copyright (c) 2021 Sony Pictures Imageworks, et al.
# SPDX-License-Identifier: Apache-2.0
# https://github.com/imageworks/spk

from typing import Any, Union
import argparse

import structlog

import spk

from . import _flags

_LOGGER = structlog.get_logger("cli")
_VALID_STAGES = ("sources", "build", "install")


def register(
    sub_parsers: argparse._SubParsersAction, **parser_args: Any
) -> argparse.ArgumentParser:

    test_cmd = sub_parsers.add_parser("test", help=_test.__doc__, **parser_args)
    test_cmd.add_argument(
        "--here",
        action="store_true",
        help=(
            "Test in the current directory, instead of the source package "
            "(mostly relevant when testing source and build stages)"
        ),
    )
    test_cmd.add_argument(
        "packages",
        metavar="FILE|PKG[@STAGE] ...",
        nargs="*",
        default=[""],
        help=(
            "The package(s) to test. Can be a file name or <name>/<version> of an existing package. "
            f"If stage is given is should be one of: {', '.join(_VALID_STAGES)}"
        ),
    )
    _flags.add_repo_flags(test_cmd, default_local=True)
    _flags.add_option_flags(test_cmd)
    _flags.add_runtime_flags(test_cmd)
    test_cmd.set_defaults(func=_test)
    return test_cmd


def _test(args: argparse.Namespace) -> None:
    """Run package tests, to run install tests the package must have been built already."""

    runtime = _flags.ensure_active_runtime(args)

    options = _flags.get_options_from_flags(args)
    repos = _flags.get_repos_from_repo_flags(args)
    for package in args.packages:
        name, *stages = package.split("@", 1)
        stages = stages or _VALID_STAGES

        try:
            spec, filename = _flags.find_package_spec(name)
        except FileNotFoundError:
            filename = package
            pkg = spk.api.parse_ident(package)
            for repo in repos.values():
                try:
                    spec = repo.read_spec(pkg)
                    break
                except spk.storage.PackageNotFoundError:
                    continue
            else:
                raise spk.storage.PackageNotFoundError(package)

        for stage in stages:
            _LOGGER.info(f"Testing {filename}@{stage}...")

            tested = set()
            for variant in spec.build.variants:

                if not args.no_host:
                    opts = spk.api.host_options()
                else:
                    opts = spk.api.OptionMap()

                opts.update(variant)
                opts.update(options)
                digest = opts.digest()
                if digest in tested:
                    continue
                tested.add(digest)

                for index, test in enumerate(spec.tests):
                    if test.stage != stage:
                        continue

                    for selector in test.selectors:
                        selected_opts = opts.copy()
                        selected_opts.update(selector)
                        if selected_opts.digest() == digest:
                            break
                    else:
                        if test.selectors:
                            _LOGGER.info(
                                "SKIP: variant not selected", test=index, variant=opts
                            )
                            continue
                    _LOGGER.info("Running test", test=index, variant=opts)

                    tester: Union[
                        spk.test.PackageSourceTester,
                        spk.test.PackageBuildTester,
                        spk.test.PackageInstallTester,
                    ]
                    if stage == "sources":
                        tester = spk.test.PackageSourceTester(spec, test.script)
                    elif stage == "build":
                        tester = spk.test.PackageBuildTester(spec, test.script)
                    elif stage == "install":
                        tester = spk.test.PackageInstallTester(spec, test.script)
                    else:
                        raise ValueError(
                            f"Untestable stage '{stage}', must be one of {_VALID_STAGES}"
                        )

                    tester = (
                        tester.with_options(opts)
                        .with_repositories(repos.values())
                        .with_requirements(test.requirements)
                    )
                    if args.here:
                        tester = tester.with_source(".")
                    try:
                        tester.test()
                    except spk.SolverError:
                        _LOGGER.error("failed to resolve test environment")
                        if args.verbose:
                            graph = tester.get_solve_graph()
                            print(
                                spk.io.format_solve_graph(graph, verbosity=args.verbose)
                            )
                        raise

File: src/cli/cmd_test.rs
// Copyright (c) 2022 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk
use std::str::FromStr;
use std::sync::Arc;

use anyhow::Result;
use clap::Args;
use spk::io::Format;
use spk::prelude::*;

use super::{flags, CommandArgs, Run};

#[cfg(test)]
#[path = "./cmd_test_test.rs"]
mod cmd_test_test;

/// Run package tests
///
/// In order to run install tests the package must have been built already
#[derive(Args)]
pub struct Test {
    #[clap(flatten)]
    pub options: flags::Options,
    #[clap(flatten)]
    pub runtime: flags::Runtime,
    #[clap(flatten)]
    pub repos: flags::Repositories,

    #[clap(short, long, global = true, parse(from_occurrences))]
    pub verbose: u32,

    #[clap(flatten)]
    pub formatter_settings: flags::DecisionFormatterSettings,

    /// Test in the current directory, instead of the source package
    ///
    /// This is mostly relevant when testing source and build stages
    #[clap(long)]
    here: bool,

    /// The package(s) to test
    ///
    /// This can be a file name or <name>/<version> of an existing package
    /// from the repository. In either case, a stage can be specified to
    /// limit which tests are executed.
    #[clap(name = "FILE|PKG[@STAGE]", required = true)]
    packages: Vec<String>,

    /// Test only the specified variant, by index, if defined
    #[clap(long, hide = true)]
    pub variant: Option<usize>,
}

#[async_trait::async_trait]
impl Run for Test {
    async fn run(&mut self) -> Result<i32> {
        let options = self.options.get_options()?;
        let (_runtime, repos) = tokio::try_join!(
            self.runtime.ensure_active_runtime(),
            self.repos.get_repos_for_non_destructive_operation()
        )?;
        let repos = repos
            .into_iter()
            .map(|(_, r)| Arc::new(r))
            .collect::<Vec<_>>();

        let source = if self.here { Some(".".into()) } else { None };

        for package in &self.packages {
            let (name, stages) = match package.split_once('@') {
                Some((name, stage)) => {
                    let stage = spk::api::TestStage::from_str(stage)?;
                    (name.to_string(), vec![stage])
                }
                None => {
                    let stages = vec![
                        spk::api::TestStage::Sources,
                        spk::api::TestStage::Build,
                        spk::api::TestStage::Install,
                    ];
                    (package.to_string(), stages)
                }
            };

            let (recipe, filename) = match flags::find_package_template(&Some(name.clone()))? {
                flags::FindPackageTemplateResult::Found { path, template } => {
                    let recipe = template.render(&options)?;
                    (Arc::new(recipe), path)
                }
                _ => {
                    let pkg = spk::api::parse_ident(&name)?;
                    let mut found = None;
                    for repo in repos.iter() {
                        match repo.read_recipe(&pkg).await {
                            Ok(recipe) => {
                                found = Some((recipe, std::path::PathBuf::from(&name)));
                                break;
                            }
                            Err(spk::Error::PackageNotFoundError(_)) => continue,
                            Err(err) => return Err(err.into()),
                        }
                    }
                    found.ok_or(spk::Error::PackageNotFoundError(pkg))?
                }
            };

            for stage in stages {
                tracing::info!("Testing {}@{stage}...", filename.display());

                let mut tested = std::collections::HashSet::new();

                let variants_to_test = match self.variant {
                    Some(index) if index < recipe.default_variants().len() => {
                        recipe.default_variants().iter().skip(index).take(1)
                    }
                    Some(index) => {
                        anyhow::bail!(
                            "--variant {index} is out of range; {} variant(s) found in {}",
                            recipe.default_variants().len(),
                            recipe.ident().format_ident(),
                        );
                    }
                    None => recipe.default_variants().iter().skip(0).take(usize::MAX),
                };

                for variant in variants_to_test {
                    let mut opts = match self.options.no_host {
                        true => spk::api::OptionMap::default(),
                        false => spk::api::host_options()?,
                    };

                    opts.extend(variant.clone());
                    opts.extend(options.clone());
                    let digest = opts.digest();
                    if !tested.insert(digest) {
                        continue;
                    }

                    for (index, test) in recipe.get_tests(&opts)?.into_iter().enumerate() {
                        if test.stage != stage {
                            continue;
                        }

                        let mut selected = false;
                        for selector in test.selectors.iter() {
                            let mut selected_opts = opts.clone();
                            selected_opts.extend(selector.clone());
                            if selected_opts.digest() == digest {
                                selected = true;
                            }
                        }
                        if !selected && !test.selectors.is_empty() {
                            tracing::info!(
                                "SKIP #{index}: variant not selected: {}",
                                spk::io::format_options(&opts)
                            );
                            continue;
                        }
                        tracing::info!(
                            "Running test #{index} variant={}",
                            spk::io::format_options(&opts)
                        );

                        let mut builder =
                            self.formatter_settings.get_formatter_builder(self.verbose);
                        let src_formatter = builder.with_header("Source Resolver ").build();
                        let build_src_formatter =
                            builder.with_header("Build Source Resolver ").build();
                        let build_formatter = builder.with_header("Build Resolver ").build();
                        let install_formatter =
                            builder.with_header("Install Env Resolver ").build();

                        match stage {
                            spk::api::TestStage::Sources => {
                                spk::test::PackageSourceTester::new(
                                    (*recipe).clone(),
                                    test.script.join("\n"),
                                )
                                .with_options(opts.clone())
                                .with_repositories(repos.iter().cloned())
                                .with_requirements(test.requirements.clone())
                                .with_source(source.clone())
                                .watch_environment_resolve(&src_formatter)
                                .test()
                                .await?
                            }

                            spk::api::TestStage::Build => {
                                spk::test::PackageBuildTester::new(
                                    (*recipe).clone(),
                                    test.script.join("\n"),
                                )
                                .with_options(opts.clone())
                                .with_repositories(repos.iter().cloned())
                                .with_requirements(test.requirements.clone())
                                .with_source(
                                    source
                                        .clone()
                                        .map(spk::build::BuildSource::LocalPath)
                                        .unwrap_or_else(|| {
                                            spk::build::BuildSource::SourcePackage(
                                                recipe
                                                    .ident()
                                                    .with_build(spk::api::Build::Source)
                                                    .into(),
                                            )
                                        }),
                                )
                                .with_source_resolver(&build_src_formatter)
                                .with_build_resolver(&build_formatter)
                                .test()
                                .await?
                            }

                            spk::api::TestStage::Install => {
                                spk::test::PackageInstallTester::new(
                                    (*recipe).clone(),
                                    test.script.join("\n"),
                                )
                                .with_options(opts.clone())
                                .with_repositories(repos.iter().cloned())
                                .with_requirements(test.requirements.clone())
                                .with_source(source.clone())
                                .watch_environment_resolve(&install_formatter)
                                .test()
                                .await?
                            }
                        }
                    }
                }
            }
        }
        Ok(0)
    }
}

impl CommandArgs for Test {
    fn get_positional_args(&self) -> Vec<String> {
        // The important positional args for a test are the packages
        self.packages.clone()
    }
}
